src/runtime_src/core/edge/drm/zocl/zocl_generic_cu.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * MPSoC based OpenCL accelerators Generic Compute Units.
+ *
+ * Copyright (C) 2020 Xilinx, Inc. All rights reserved.
+ *
+ * Authors:
+ *    Min Ma      <min.ma@xilinx.com>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <drm/drmP.h>
+#include <linux/anon_inodes.h>
+#include "zocl_util.h"
+#include "sched_exec.h"
+#include "zocl_generic_cu.h"
+
+static int generic_cu_release(struct inode *inode, struct file *filp)
+{
+	struct generic_cu *gcu = filp->private_data;
+	struct generic_cu_info *info = gcu->info;
+	int ret;
+
+	free_irq(info->irq, gcu);
+	ret = sched_attach_cu(gcu->zdev, info->cu_idx);
+	if (ret < 0)
+		DRM_WARN("Scheduler attach CU[%d] failed.\n", info->cu_idx);
+
+	kfree(info->name);
+	kfree(info);
+	kfree(gcu);
+	return 0;
+}
+
+static __poll_t generic_cu_poll(struct file *filp, poll_table *wait)
+{
+	struct generic_cu *gcu = filp->private_data;
+	__poll_t ret = 0;
+
+	poll_wait(filp, &gcu->waitq, wait);
+
+	/* Safe to read only one object without lock */
+	if (gcu->event > 0)
+		ret = POLLIN;
+
+	return ret;
+}
+
+static ssize_t generic_cu_read(struct file *filp, char __user *buf,
```

Comment:
the implementation of blocking and waiting looks quite complicated here. i'm wondering if you should just call wait_event_interruptible()? (see mailbox_read() for example).

Reply:
Thanks. Done.